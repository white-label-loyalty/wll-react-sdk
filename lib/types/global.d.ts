// Global type definitions for React Native environment

declare const __DEV__: boolean;
