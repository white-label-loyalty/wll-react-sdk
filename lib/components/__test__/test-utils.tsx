import { render } from '@testing-library/react';
import React from 'react';
import { CoreApiUrl, WllSdkProvider } from '../../context/WllSdkContext';
import { defaultTheme } from '../../utils/styling';

const defaultConfig = {
  apiKey: 'test-api-key',
  userToken: 'test-user-token',
  coreApiUrl: CoreApiUrl.STAGING_EU,
};

const AllTheProviders = ({ children }: { children: React.ReactNode }) => {
  return (
    <WllSdkProvider
      config={defaultConfig}
      theme={defaultTheme}
      navigationConfig={{}}
    >
      {children}
    </WllSdkProvider>
  );
};

const customRender = (ui: React.ReactElement, options = {}) =>
  render(ui, { wrapper: AllTheProviders, ...options });

// re-export everything
export * from '@testing-library/react';

// override render method
export { customRender as render };
