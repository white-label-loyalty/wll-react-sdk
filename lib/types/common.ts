import { ImageProps, ViewStyle } from 'react-native';

export type ImagePropsNoSource = Omit<ImageProps, 'source'>;

export type FlexDirection = 'row' | 'row-reverse' | 'column' | 'column-reverse';

export type Justify =
  | 'start'
  | 'end'
  | 'center'
  | 'between'
  | 'around'
  | 'evenly';

export type Align = 'start' | 'end' | 'center' | 'stretch' | 'baseline';

export type LayoutProps = {
  children: React.ReactNode;
  justify?: Justify;
  align?: Align;
  direction?: FlexDirection;
  style?: ViewStyle;
};

export const justifyMap: Record<Justify, ViewStyle['justifyContent']> = {
  start: 'flex-start',
  end: 'flex-end',
  center: 'center',
  between: 'space-between',
  around: 'space-around',
  evenly: 'space-evenly',
} as const;

export const alignMap: Record<Align, ViewStyle['alignItems']> = {
  start: 'flex-start',
  end: 'flex-end',
  center: 'center',
  stretch: 'stretch',
  baseline: 'baseline',
} as const;
