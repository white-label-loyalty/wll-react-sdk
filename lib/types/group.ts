import { TSection } from './section';

export type TGroup = {
  name: string;
  active: boolean;
  id: string;
  createdAt: string;
  updatedAt: string;
  sections: TSection[];
};
