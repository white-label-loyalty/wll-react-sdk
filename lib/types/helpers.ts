/**
 * Makes a configuration type with required and optional fields,
 * providing default values for optional fields.
 */
export type SafeConfig<T, R extends keyof T> = Required<Pick<T, R>> &
  Partial<Omit<T, R>>;

/**
 * Adds optional children to any component props.
 */
export type WithChildren<T = {}> = T & { children?: React.ReactNode };
