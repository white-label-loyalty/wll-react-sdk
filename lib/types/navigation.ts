export type NavigationType = 'external' | 'internal' | 'modal';

type NavigationHandlerParams = {
  target: string;
  windowTarget: string;
};

export type NavigationConfig = {
  navigationHandlers?: {
    external?: (params: NavigationHandlerParams) => Promise<void>;
    internal?: (params: NavigationHandlerParams) => Promise<void>;
    modal?: (params: NavigationHandlerParams) => Promise<void>;
  };
  external?: (params: NavigationHandlerParams) => Promise<void>;
  internal?: (params: NavigationHandlerParams) => Promise<void>;
  modal?: (params: NavigationHandlerParams) => Promise<void>;
  baseUrl?: string;
};
