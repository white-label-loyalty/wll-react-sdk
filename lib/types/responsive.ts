export type DimensionMode = 'desktop' | 'tablet' | 'mobile';

export type WindowDimensions = {
  width: number;
  height: number;
};

export type ResponsiveContextType = {
  dimensionMode: DimensionMode;
  isTablet: boolean;
  isMobile: boolean;
  isDesktop: boolean;
};
