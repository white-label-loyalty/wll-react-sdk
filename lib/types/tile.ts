export enum CtaAction {
  sameWindow = 'SAME_WINDOW',
  newWindow = 'NEW_WINDOW',
}

export enum BadgeTileType {
  latestEarned = 'LATEST_EARNED',
  specificBadge = 'SPECIFIC_BADGE',
}

export enum TierTileType {
  currentTier = 'CURRENT_TIER',
  nextTier = 'NEXT_TIER',
  specificTier = 'SPECIFIC_TIER',
}

export enum TileType {
  Banner = 'BANNER',
  Points = 'POINTS',
  Content = 'CONTENT',
  Reward = 'REWARD',
  Badge = 'BADGE',
  RewardCategory = 'REWARD_CATEGORY',
  Tier = 'TIER',
}

export enum TileHeight {
  Half = 'HALF',
  Full = 'FULL',
}
type Badge = {
  id: string | null;
  name: string | null;
  description: string | null;
  artworkUrl: string | null;
  createdAt: string | null;
  updatedAt: string | null;
};
export class RewardCategory {
  name?: string | null;
  priority?: number;
  type?: string | null;
  id?: string | null;
  createdAt?: string | null;
  updatedAt?: string | null;
  description?: string | null;
  metadata?: string | null;
  pictureUrl?: string | null;
  rewards?: Reward[];
  parent?: string | null;
}

export class Reward {
  id?: string | null;
  createdAt?: string | null;
  updatedAt?: string | null;
  name?: string | null;
  pictureUrl?: string | null;
  description?: string | null;
  value?: number | null;
  price?: number;
  priority?: number;
  discounts?: any[] | null;
  redemptionChannels?: string[] | null;
  venues?: any[] | null;
  purchasableForAudiences?: string[] | null;
  category?: string | null;
  terms?: string | null;
  availability?: Availability;
  purchasable?: boolean;
  tier?: string | null;
  summary?: string | null;
  redemptionMessage?: string | null;
  visibilityCriteria?: string | null;
  logoUrl?: string | null;
  type?: string | null;
  codeType?: string | null;
  code?: string | null;
  purchaseExpiration?: string | null;
  hideCode?: boolean;
  notificationConfig?: string | null;
}
type Availability = {
  start: string | null;
  end: string | null;
};
type Effectivity = {
  start: string | null;
  end: string | null;
};

export type TierType = {
  id: string | null;
  name: string | null;
  description: string | null;
  artworkUrl: string | null;
  priority: number;
  pointsRequirement: number;
  calculation: string | null;
  accumulationPeriod: string | null;
  effectivity: Effectivity;
  createdAt: string | null;
  updatedAt: string | null;
  earnedPoints: number;
  attained: boolean;
};

export class BannerTileConfig {
  imageUrl?: string | null;
  title?: string | null;
  description?: string | null;
  ctaText?: string | null;
  ctaLink?: string | null;
  ctaAction?: CtaAction;
}

export class PointsTileConfig {
  title?: string | null;
  multiplier?: number;
  prefix?: string | null;
  suffix?: string | null;
  imageUrl?: string | null;
  points?: number;
}

export class ContentTileConfig {
  title?: string | null;
  subtitle?: string | null;
  imageUrl?: string | null;
}

export class RewardTileConfig {
  reward?: Reward;
  showPrice?: boolean;
}

export class BadgeTileConfig {
  badgeTileType?: BadgeTileType;
  badgeId?: string | null;
  badge?: Badge;
}

export class RewardCategoryTileConfig {
  categoryId?: string | null;
  allowDecorationOverlay?: boolean;
  rewardCategory?: RewardCategory;
}

export class TierTileConfig {
  tierId?: string | null;
  tier?: TierType;
  type?: TierTileType;
}

export type Tile = {
  id: string | null;
  type: TileType;
  active: boolean;
  createdAt: string | null;
  updatedAt: string | null;
  visibilityCriteria: object | null;
  tileHeight: TileHeight;
  configuration: TileConfig;
};

const getConfigForTileType = (tileType: TileType) => {
  switch (tileType) {
    case TileType.Banner:
      return BannerTileConfig;
    case TileType.Points:
      return PointsTileConfig;
    case TileType.Content:
      return ContentTileConfig;
    case TileType.Reward:
      return RewardTileConfig;
    case TileType.Badge:
      return BadgeTileConfig;
    case TileType.RewardCategory:
      return RewardCategoryTileConfig;
    case TileType.Tier:
      return TierTileConfig;
    default:
      throw new Error(`Tile type ${tileType} is not supported`);
  }
};

export type TileConfig = InstanceType<
  ReturnType<typeof getConfigForTileType>
>;
