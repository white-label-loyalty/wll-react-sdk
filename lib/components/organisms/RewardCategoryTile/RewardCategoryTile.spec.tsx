import '@testing-library/jest-dom';
import { screen } from '@testing-library/react';
import React from 'react';
import { TileHeight, TileType } from '../../../types/tile';
import { render } from '../../__test__/test-utils';
import { RewardCategoryTile } from './index';

export const RewardCategoryTileMock = {
  tileHeight: TileHeight.Full,
  active: true,
  type: TileType.RewardCategory,
  configuration: {
    showName: true,
    rewardCategoryId: '9d8287e0-bae9-47b4-8bf0-0115fd64336a',
    artworkUrl:
      'https://images.pexels.com/photos/1850595/pexels-photo-1850595.jpeg?auto=compress&cs=tinysrgb&w=1260&h=750&dpr=2',
    name: 'Restaurant Rewards',
  },
  id: 'f57f2f98-cabf-4b81-82ac-a0249172d65c',
  createdAt: '2024-10-18T14:40:37.835Z',
  updatedAt: '2024-10-18T14:40:37.835Z',
  priority: 8,
};

describe('<RewardCategoryTile />', () => {
  it('renders and matches snapshot', () => {
    const { container } = render(
      <RewardCategoryTile tile={RewardCategoryTileMock} />
    );
    expect(container).toMatchSnapshot();
  });
});

describe('<RewardCategoryTile /> Content', () => {
  it('renders category name when showName is true', () => {
    render(<RewardCategoryTile tile={RewardCategoryTileMock} />);
    const rewardCategoryHeader = screen.getByTestId('reward-category-header');
    expect(rewardCategoryHeader.textContent).toBe('Restaurant Rewards');
  });

  it('does not render category name when showName is false', () => {
    const mockWithoutName = {
      ...RewardCategoryTileMock,
      configuration: {
        ...RewardCategoryTileMock.configuration,
        showName: false,
      },
    };
    const { queryByTestId } = render(
      <RewardCategoryTile tile={mockWithoutName} />
    );
    const rewardCategoryHeader = queryByTestId('reward-category-header');
    expect(rewardCategoryHeader).not.toBeInTheDocument();
  });
});

describe('<RewardCategoryTile /> Media', () => {
  it('renders the artwork image', () => {
    const { getByTestId } = render(
      <RewardCategoryTile tile={RewardCategoryTileMock} />
    );
    expect(getByTestId('reward-category-media')).toBeInTheDocument();
  });

  it('handles missing artworkUrl gracefully', () => {
    const mockWithoutArtwork = {
      ...RewardCategoryTileMock,
      configuration: {
        ...RewardCategoryTileMock.configuration,
        artworkUrl: '',
      },
    };
    const { queryByTestId } = render(
      <RewardCategoryTile tile={mockWithoutArtwork} />
    );
    expect(queryByTestId('reward-category-media')).not.toBeInTheDocument();
  });
});

describe('<RewardCategoryTile /> Edge Cases', () => {
  it('handles missing name gracefully', () => {
    const mockWithoutName = {
      ...RewardCategoryTileMock,
      configuration: {
        ...RewardCategoryTileMock.configuration,
        name: undefined,
      },
    };
    const { queryByTestId } = render(
      <RewardCategoryTile tile={mockWithoutName} />
    );
    expect(queryByTestId('reward-category-header')).not.toBeInTheDocument();
  });

  it('handles inactive tile gracefully', () => {
    const mockInactiveTile = {
      ...RewardCategoryTileMock,
      active: false,
    };
    const { container } = render(
      <RewardCategoryTile tile={mockInactiveTile} />
    );
    expect(container.firstChild).toBeNull();
  });
});
