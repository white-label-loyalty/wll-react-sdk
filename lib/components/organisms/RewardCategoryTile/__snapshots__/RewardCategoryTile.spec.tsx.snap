// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RewardCategoryTile /> renders and matches snapshot 1`] = `
<div>
  <button
    aria-label="Tile"
    class="css-view-175oi2r r-cursor-1loqt21 r-touchAction-1otgn73 r-aspectRatio-1ujkv8a r-borderRadius-y47klf r-height-1pi2tsx r-overflow-1udh08x r-position-bnwqim r-width-13qz1uu"
    role="button"
    style="background-color: rgb(248, 247, 252); opacity: 1; flex-direction: column; justify-content: flex-start; align-items: stretch;"
    tabindex="0"
    type="button"
  >
    <h1
      aria-label="Reward category: Restaurant Rewards"
      class="css-view-175oi2r r-alignItems-1awozwy r-height-17sp8yy r-justifyContent-1777fci r-left-1d2f490 r-position-u8s1d r-right-zchlnj r-top-ipm5af r-zIndex-mhe3cw"
      data-testid="reward-category-header"
      role="heading"
      style="background-color: rgb(57, 46, 215);"
    >
      <div
        class="css-text-146c3p1 r-maxWidth-dnmrzs r-overflow-1udh08x r-textOverflow-1udbk01 r-whiteSpace-3s2u2q r-wordWrap-1iln25a r-fontSize-ubezar r-paddingInline-1rds9hb"
        dir="auto"
        style="font-family: -apple-system,BlinkMacSystemFont,Segoe UI,Roboto,Oxygen-Sans,Ubuntu,Cantarell,Helvetica Neue,sans-serif; color: rgb(255, 255, 255);"
      >
        Restaurant Rewards
      </div>
    </h1>
    <div
      class="css-view-175oi2r r-alignItems-1awozwy r-justifyContent-1777fci r-height-1pi2tsx r-objectFit-1t7uo4s r-position-u8s1d r-width-13qz1uu"
      style="background-color: rgba(57, 46, 215, 0.2);"
    >
      <div
        class="css-view-175oi2r r-bottom-1p0dtai r-left-1d2f490 r-position-u8s1d r-right-zchlnj r-top-ipm5af"
        style="background-color: rgba(57, 46, 215, 0.2); opacity: 1;"
      />
      <div
        aria-label="Reward category image for Restaurant Rewards"
        class="css-view-175oi2r r-flexBasis-1mlwlqe r-overflow-1udh08x r-zIndex-417010 r-bottom-1p0dtai r-left-1d2f490 r-position-u8s1d r-right-zchlnj r-top-ipm5af"
        data-testid="reward-category-media"
        role="img"
        style="opacity: 0;"
      >
        <div
          class="css-view-175oi2r r-backgroundColor-1niwhzg r-backgroundPosition-vvn4in r-backgroundRepeat-u6sd8q r-bottom-1p0dtai r-height-1pi2tsx r-left-1d2f490 r-position-u8s1d r-right-zchlnj r-top-ipm5af r-width-13qz1uu r-zIndex-1wyyakw r-backgroundSize-4gszlv"
          style="background-image: url(https://images.pexels.com/photos/1850595/pexels-photo-1850595.jpeg?auto=compress&cs=tinysrgb&w=1260&h=750&dpr=2);"
        />
        <img
          alt="Reward category image for Restaurant Rewards"
          class="css-accessibilityImage-9pa8cd"
          draggable="false"
          src="https://images.pexels.com/photos/1850595/pexels-photo-1850595.jpeg?auto=compress&cs=tinysrgb&w=1260&h=750&dpr=2"
        />
      </div>
    </div>
  </button>
</div>
`;
